### ISSUE TITLE
fix link local IPv6 interface detection twisted/internet/test/ for non-OSX BSD

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6843]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)| @oberstet reported|
|-|-|
|Trac ID|trac#6843|
|Type|enhancement|
|Created|2013-11-18 19:05:21Z|
|Branch|https://github.com/twisted/twisted/tree/freebsd-posixifaces-6843|

test_tcp does specify wrong struct via ctypes. this should be

```
struct sockaddr_in {
	uint8_t		sin_len;
	sa_family_t	sin_family;
	in_port_t	sin_port;
	struct	in_addr sin_addr;
	char	sin_zero[8];
};
```

Attachments:

* [fix_6843.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f24/f244914c4a922b638f68af26f75213391cffc777/e2ec936c6dce0a1f51c51f30349e2cd9c227d010.patch) (554 bytes) - added by oberstet on 2013-11-18 19:10:51Z - 
* [fix_6843.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f24/f244914c4a922b638f68af26f75213391cffc777/782c914d78deb8af131c72cb1f0f0e07eef57769.patch) (554 bytes) - added by oberstet on 2013-11-18 19:20:51Z - this one .. I mixed up whats being diffed against what
* [fix_6843.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f24/f244914c4a922b638f68af26f75213391cffc777/1fae45ac7a46d0ccf2ca59b93fde9bec533d88b2.patch) (554 bytes) - added by oberstet on 2013-11-18 20:01:05Z - ok, only do stuff for FreeBSD, not the other BSDs .. no test systems
* [fix_6843.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f24/f244914c4a922b638f68af26f75213391cffc777/8ed313d58de03925e9fbf47a06a6835d4cd5124c.patch) (460 bytes) - added by oberstet on 2013-11-18 20:04:04Z - now ..
* [fix_6843.5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f24/f244914c4a922b638f68af26f75213391cffc777/3d240cfb29662bd4b125d9a874a7eaddb120025a.patch) (1029 bytes) - added by oberstet on 2013-11-18 23:52:04Z - kill embedded interface indicies in scoped link-local addresses
* [fix_6843.6.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f24/f244914c4a922b638f68af26f75213391cffc777/11c95164f4d595e9ef258a16399af79f2e32efd6.patch) (1164 bytes) - added by oberstet on 2013-11-19 13:30:16Z - guard both changes for "freebsd"

<details><summary>Searchable metadata</summary>

```
trac-id__6843 6843
type__enhancement enhancement
reporter__oberstet oberstet
priority__normal normal
milestone__ 
branch__branches_freebsd_posixifaces_6843 branches/freebsd-posixifaces-6843
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1384801521000000 1384801521000000
changetime__1384947812000000 1384947812000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Hm, thought I filed this already but I can't find a ticket...

This is probably most easily fixed using cffi.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Attached patch is backwards, fwiw :)  Also we have no OpenBSD or NetBSD build slaves to test those parts of the patch on.

fwiw, I suggested cffi because it would remove all of this extra checking (but clearly the patch would be longer).

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_5"></a>|@oberstet commented|
|-|-|

Not sure: are you suggesting to rewrite the _whole_ thing from ctypes to cffi to get link local IPv6 detection correct in the tests?

Mmmh. As much as I agree that cffi is the way, and ctypes sucks, I'm not sure: the fix is essentially a 1 liner ..

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_6"></a>|@oberstet commented|
|-|-|

Btw: are you still around on IRC .. this "you can't save because someone else edited" Trac shit sucks a little for interactive communication..

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Sure, I'm usually around in #twisted-dev.

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)|@oberstet set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [40780]) Branching to 'freebsd-posixifaces-6843'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @oberstet|
|-|-|

Thanks!  I applied the patch to a branch and did a build.  The tests that used to be skipped now aren't skipped anymore.  But many of them fail with errors kind of like this one:

```
a = IPv6Address(TCP, 'fe80::a00:27ff:fed9:bbe9%em0', 18434)
b = IPv6Address(TCP, 'fe80:1::a00:27ff:fed9:bbe9%em0', 18434)
```

where `a` and `b` were expected to be equal to each other.

Here's a link to the problem output of one such build - https://buildbot.twistedmatrix.com/builders/freebsd-9.2-amd64-python2.7/builds/1/steps/select/logs/problems

This is a good step towards working better on FreeBSD.  Any idea how we can get the tests to actually pass though?

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)<a name="note_11"></a>|@oberstet commented|
|-|-|

Let me just link a couple of resources that were discussed on IRC:

http://www.freebsd.org/doc/en/books/developers-handbook/ipv6.html#ipv6-scope-index
http://www.freebsd.org/cgi/man.cgi?query=getaddrinfo&sektion=3
https://gist.github.com/oberstet/7535218
https://gist.github.com/oberstet/7535412
https://gist.github.com/oberstet/7535418
http://www.freebsd.org/cgi/man.cgi?ifmib

**Comment by twisted-trac** [comment]
|[<img alt="oberstet's avatar" src="https://avatars.githubusercontent.com/u/233340?s=50" width="50" height="50">](https://github.com/oberstet)|@oberstet set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun removed owner|
|-|-|

Checked the latest patch in.  There was an unfortunate bug that caused it to produce illegal IPv6 addresses.  I fixed this too.  I think the branch looks good now.  Someone else should give it a check, though, since I wrote part of the code in it now.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to @lvh|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_15"></a>|@lvh set owner to @exarkun|
|-|-|

Hi JP. Thanks for working on this.

- There appear to be some [remaining twistedchecker errors](https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/1531/steps/run-twistedchecker/logs/new%20twistedchecker%20errors). They appear to be in the tests, which you didn't appear to touch. Since it's a tiny change, perhaps it would be good to opportunistically do it here. However, that brings me to my next point.
- This doesn't appear to touch the tests. From what I understand, there were previously skipped/failing tests that now work. Those *seem* to be reliable to me. I suppose that as long as we have a buildbot for each supported platform (specifically both FreeBSD 10, and 9, I suppose, since 9 needs this behavior and 10 doesn't), and the behavior of the platforms does not change, this should not be a problem :)
- And, perhaps a bit bikesheddy, but I had to ask: why I{} instead of C{} for socket.AF_* and in_addr? I don't think this has to be changed, FWIW; it's more for future work I do on twisted :)

Thanks again for your contribution!

cheers

lvh

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

> There appear to be some  remaining twistedchecker errors. They appear to be in the tests, which you didn't appear to touch.

Neither of those errors makes any sense to me. :(  The line being blamed isn't even a comment.  Oh, I see.  The URL has a *fragment*.  Uuuuuuuuuuugh.  twistedchecker bug.

> This doesn't appear to touch the tests. From what I understand, there were previously skipped/failing tests that now work.

Indeed.  I don't know what the condition of IPv6 support was on the recently retired FreeBSD !~8 slave was (I don't think old builds are visible once the builder is removed).  But this fixes a number of failing tests on the new FreeBSD 9 slave.

> specifically both FreeBSD 10, and 9, I suppose, since 9 needs this behavior and 10 doesn't

There isn't a FreeBSD 10 slave though (I don't think FreeBSD 10 is out yet but I dunno).  The comment is a hint to a future maintainer who thinks the code is nonsense because no version of FreeBSD they've ever seen requires this work-around.  It lets them know that if they go waaaaaaay back to FreeBSD 9 they'll see the system behavior being worked around.  Lack of a FreeBSD 10 slave is the reason I put in a comment instead of an implementation switch. :)

> And, perhaps a bit bikesheddy, but I had to ask: why I{} instead of C{} for socket.AF_* and in_addr?

I didn't use `C{}` because `socket.AF_*` is a syntax error. :)  I use `I{}` as a catch-all for "this is special in some way, not just some junky English words".  I'm not sure if this is a good idea or not but it seems to beat not using any markup at all.  If anything, I could perhaps imagine teaching pydoctor to support `L{socket.AF_*}` (or something like it) to mean "generate a bunch of links to all of the things you can find that match this pattern".  You'd get like 80 jillion annotations on the rendered text (maybe a bunch of super- or sub-script endnote-style things) and maybe that'd be okay.  On the other hand, it'd be much better to be able to write `L{socket.AddressFamilyType}`.  Too bad that's not a thing (the first thing I typed was `L{int}` but then I deleted it because that's stupid and useless).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_17"></a>|@lvh commented|
|-|-|

Heh, okay; cool. Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set status to `closed`|
|-|-|

(In [40810]) Merge freebsd-posixifaces-6843 

Author: oberstet, exarkun
Reviewer: exarkun, lvh
Fixes: [#6843](https://github.com/twisted/twisted/issues/6843)

Account for the embedded scope id/interface index value found in the results of
`getifaddrs` on current versions of FreeBSD (by discarding them).  This fixes
seemingly incorrect link local address information retrieved by the IPv6 testing
helper.
